### ISSUE TITLE
Feature Request: directory shortcuts

### ISSUE THREAD

**Issue by felipec** [fish-shell/fish-shell#7493]
As stated in issue #1449, zsh has a great feature that many can't live without: directory aliases.

Directory aliases are just like normal aliases, except they apply only to directories, and can be expanding by using tilde (`~`), just like home directories.

For example:

```sh
hash -d d=~/dev
hash -d mnt=/run/media/$USER
hash -d dotfiles=~/.config/dotfiles
```

This has a number of advantages:

### 1. It's easy to `cd`

Obviously it's much easier to type `cd ~mnt` than `cd /run/media/$USER`.

One could argue that could also be achieved with a variable and simply type `cd $mnt`, however that would not autocomplete the following directories correctly, like `cd $mnt/back<tab>`.

Additionally, `cd $m<tab>` could autocomplete undesired variables.

This alone justifies the feature, which was not explored correctly in issue #1449.

### 2. It gives a nicer prompt

The fish function `prompt_pwd` gives a nice shortcut, like "/r/m/felipec", however, that's not short enough, nor descriptive enough, certainly not like what zsh's prompt expansion `%~` gives: "~mnt".

I have written tools that allow me to track in which directories I type commands, and as result I know 20% of my commands I type in my home directory (`~`), but 10% I do on "/srv/http/project". I would like to see "~project", not "/s/h/project".

Moreover, I would like `cd (prompt_pwd)` to work: `cd ~project`.

### 3. It gives a nicer title

This is basically the same as 2, except for the title of the window.

---

I have used zsh for many years, and now I take this feature for granted, and even though fish looks really nice I'm sad to say without this feature it's not a viable replacement.

**Comment by faho** [comment]
No, thank you, as explained in #1449

**Comment by felipec** [comment]
> No, thank you, as explained in #1449

My rationale was not given in #1449, therefore no explanation against my rationale could have been given in #1449, and it wasn't.

> This is a poweruser feature that might save you 0.1s each day after you spend five hours configuring it. That's not the sort of thing we tend to add.

Directories are not a poweruser feature. A quick search shows I have used these aliases 1,500 times since the last time I cleared my history, so it's more like 1 per hour, at least.

And I just timed how much it takes to configure them in zsh, it took me 1.4 minutes, **not** five hours.

Not to mention it's not just about typing commands, which I already explained.

> If you really want it, it's possible to emulate by writing a function that does the lookup, that you then use in `cd`, `prompt_pwd` and `fish_title`. It's not something that needs to be in fish proper.

Yes, how about the rest of the commands?

The last one I used was:

    vim ~c/zsh/Completion/Unix/Command/_git

And no thank you. The whole point of fish is what it works out-of-the-box, I'm not going to hack it just to provide what I consider basic functionality.

> > I have used zsh for many years, and now I take this feature for granted, and even though fish looks really nice I'm sad to say without this feature it's not a viable replacement.
> 
> Fish isn't zsh.

Nor is it a useful shell.

It took me about half an hour writing this feature request, which explains why zsh users would never move to fish, and I bet you didn't spend more than thirty seconds writing the reply, which will ensure less fish users.

I spent about one day configuring fish, and trying it out. Thanks for wasting my time. Uninstalling.
